Backend/utils/cronJobRecommendation.js

```diff
@@ -24,50 +38,70 @@ async function calculateRecommendations() {
     for (const song of songs) {
         const { userId, genre, mood, tempo } = song;
         if (!userTagHistory[userId]) {
-            userTagHistory[userId] = [];
+            userTagHistory[userId] = []; // Initialize user's tag history if not already
         }
 
         // Assign numerical values to tags
-        const tagCombination = [
-            tagValues.genre[genre],
-            mood ? tagValues.mood[mood] : 0,
-            tempo ? tagValues.tempo[tempo] : 0
-        ];
+        const tagCombination = {
+            genre: tagValues.genre[genre], // Convert genre to number
+            mood: mood ? tagValues.mood[mood] : 0, // Convert mood to number, if exists
+            tempo: tempo ? tagValues.tempo[tempo] : 0 // Convert tempo to number, if exists
+        };
 
-        userTagHistory[userId].push(tagCombination.join('-'));
+        userTagHistory[userId].push(tagCombination); // Add tag combination to user's history
     }
 
     // Calculate vector similarity between users based on their tag histories
     const userSimilarities = {};
+    const threshold = 1.0; // Minimum threshold for recommending an artist
 
-    const users = Object.keys(userTagHistory);
+    const users = Object.keys(userTagHistory); // Get list of all users
 
+    // Compare every user with every other user
     for (let i = 0; i < users.length; i++) {
         for (let j = i + 1; j < users.length; j++) {
             const user1 = users[i];
             const user2 = users[j];
-            const commonTags = userTagHistory[user1].filter(tag => userTagHistory[user2].includes(tag));
 
-            if (commonTags.length > 0) {
-                if (!userSimilarities[user1]) {
-                    userSimilarities[user1] = [];
-                }
-                if (!userSimilarities[user2]) {
-                    userSimilarities[user2] = [];
+            // Create pairs of tags from user1 and user2 to compare them
+            const tagPairs = userTagHistory[user1].flatMap(tag1 => userTagHistory[user2].map(tag2 => ({ tag1, tag2 })));
+
+            for (const { tag1, tag2 } of tagPairs) { //For each pair of tags
+                let similarityScore = 0; // Initialize similarity score
+
+                // Compare the genre tags and add to the similarity score if they match
+                if (tag1.genre === tag2.genre) similarityScore += tagWeights.genre;
+                // Compare the mood tags and add to the similarity score if they match
+                if (tag1.mood === tag2.mood) similarityScore += tagWeights.mood;
+                // Compare the tempo tags and add to the similarity score if they match
+                if (tag1.tempo === tag2.tempo) similarityScore += tagWeights.tempo;
+
+                // If similarity score is above or equal to the threshold, save the similarity
+                if (similarityScore >= threshold) {
```

Comment:
isn't the max threshold 1.0 because 0.5 + 0.3 + 0.2 = 1.0? curious why you are checking >= and not just ==

Reply:
I've set the threshold at 0.75 for now to ensure I capture at least 2 similar tags effectively. However, I've been working on a more adaptable method for determining this threshold, which I plan to introduce in the next PR.

Reply:
gotcha! not sure if you updated it to 0.75, I still see line 56 `const threshold = 1.0;`

Reply:
Yes, I haven't pushed the changes yet